core/translate/main_loop.rs

```diff
@@ -743,23 +773,205 @@ pub fn open_loop(
+                program.emit_insn(Insn::HashBuild {
+                    cursor_id: build_cursor_id,
+                    key_start_reg: build_key_start_reg,
+                    num_keys,
+                    hash_table_reg,
+                    mem_budget: hash_join_op.mem_budget,
+                    collations,
+                });
+
+                program.preassign_label_to_next_insn(skip_to_next);
+                program.emit_insn(Insn::Next {
+                    cursor_id: build_cursor_id,
+                    pc_if_next: build_loop_start,
+                });
+
+                program.preassign_label_to_next_insn(build_loop_end);
+                program.emit_insn(Insn::HashBuildFinalize { hash_table_reg });
+
+                // Avoid unresolved labels for the build table's normal loop labels since
+                // we bypass the regular open/close loop plumbing for hash build.
+                let build_loop_labels = t_ctx.labels_main_loop[hash_join_op.build_table_idx];
+                program.preassign_label_to_next_insn(build_loop_labels.next);
+                program.preassign_label_to_next_insn(build_loop_labels.loop_end);
+
+                // Hash Table Probe Phase
+                // Iterate through probe table and look up matches in hash table
+                let probe_cursor_id = table_cursor_id.expect("Probe table must have a cursor");
+                program.emit_insn(Insn::Rewind {
+                    cursor_id: probe_cursor_id,
+                    pc_if_empty: loop_end,
+                });
+                program.preassign_label_to_next_insn(loop_start);
+
+                // Extract probe keys from current row
+                let probe_key_start_reg = program.alloc_registers(num_keys);
+                for (idx, (_build_col, probe_col)) in
+                    hash_join_op.join_key_columns.iter().enumerate()
+                {
+                    program.emit_column_or_rowid(
+                        probe_cursor_id,
+                        *probe_col,
+                        probe_key_start_reg + idx,
+                    );
+                }
+
+                // Probe hash table with keys, store matched rowid in register
+                let match_reg = program.alloc_register();
+                program.emit_insn(Insn::HashProbe {
+                    hash_table_reg,
+                    key_start_reg: probe_key_start_reg,
+                    num_keys,
+                    dest_reg: match_reg,
+                    target_pc: next, // Jump if no match
+                });
+
+                // Label for match processing, HashNext jumps here to avoid re-probing
+                let match_found_label = program.allocate_label();
+                program.preassign_label_to_next_insn(match_found_label);
+
+                // Seek build table cursor to the matched rowid
+                // The match_reg contains the rowid from the hash table
+                program.emit_insn(Insn::SeekRowid {
```

Comment:
Did you consider storing the row data in the hash table itself so that we don't incur btree probes for every match? Where is the performance gain from hash join vs ephemeral btree coming from? Is building the ephemeral table really so much slower in comparison in tursodb that we can both build the hash table AND do btree probes for every match and be faster?